pkg/ipmi/ipmi.go

```diff
@@ -173,3 +191,94 @@ func (i *IPMI) ShutoffWatchdog() error {
 
 	return nil
 }
+
+func (i *IPMI) waitForSetInProgressCleared() error {
+	req := &req{}
+	req.msg.cmd = _GET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	var data [4]byte
+
+	data[0] = 0
+	data[1] = _SET_IN_PROGRESS
+	data[2] = 0
+	data[3] = 0
+	req.msg.data = unsafe.Pointer(&data)
+	req.msg.dataLen = 4
+
+	retry := 20
+	for retry > 0 {
+		recv, err := i.sendrecv(req)
+		if err != nil {
+			return err
+		}
+		// response data byte 3 bit[1:0] == 00b indicates set complete
+		if len(recv) == 3 && (recv[2]&0x3) == 0 {
+			return nil
+		}
+		time.Sleep(1 * time.Millisecond)
+		retry--
+	}
+
+	return fmt.Errorf("Wait for Set in Progress cleared timeout")
+
+}
+
+func (i *IPMI) setsysinfo(data *setSystemInfoReq, dataLen uint16) error {
+	req := &req{}
+	req.msg.cmd = _SET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	req.msg.dataLen = dataLen
+	req.msg.data = unsafe.Pointer(data)
+
+	if err := i.waitForSetInProgressCleared(); err != nil {
```

Comment:
Is this actually needed? While checking the `ipmitool` behavior, it doesn't look like it waits for this.

Reply:
I wouldn't add this retry and wait logic. I would simply return an error here if there is any.

Reply:
This is from the IPMI spec for parameter "Set in progress", from the spec "it is used to indicate when any of the following parameters are being updated, and when the updates are completed. This bit is primarily provided to alert software that some other software or utility is in the process of making changes to the data". The UEFI BIOS employs this check so In my opinion it's safer to follow. I check double check with BMC guys if this in fact can be removed.

Reply:
I would log the error instead of having a wait and retry logic here.

Reply:
I'm still checking with our BMC guys on this.

Reply:
This depends on the BMC's implementation, although it's in the IPMI spec but our BMC doesn't process this flag so I removed it for this case. I think if it's needed to be added back it can be added to the new function and keep the setsysinfo() without the check.